lib/internal/http2/core.js

```diff
@@ -878,107 +963,146 @@ class Http2Session extends EventEmitter {
 
     state.pendingAck++;
     const settingsFn = submitSettings.bind(this, settings);
-    if (state.connecting) {
+    if (this.connecting) {
       this.once('connect', settingsFn);
       return;
     }
     settingsFn();
   }
 
-  // Destroy the Http2Session
-  destroy() {
-    const state = this[kState];
-    if (state.destroyed || state.destroying)
-      return;
-    debug(`Http2Session ${sessionName(this[kType])}: destroying`);
-    state.destroying = true;
-    state.destroyed = false;
-
-    // Unenroll the timer
-    this.setTimeout(0, sessionOnTimeout);
-
-    // Shut down any still open streams
-    const streams = state.streams;
-    streams.forEach((stream) => stream.destroy());
-
-    // Disassociate from the socket and server
-    const socket = this[kSocket];
-    // socket.pause();
-    delete this[kProxySocket];
-    delete this[kSocket];
-    delete this[kServer];
+  // Sumits a GOAWAY frame to be sent to the remote peer. Note that this
+  // is only a notification, and does not affect the usable state of the
+  // session with the notable exception that new incoming streams will
+  // be rejected automatically.
+  goaway(code = NGHTTP2_NO_ERROR, lastStreamID = 0, opaqueData) {
+    if (this.destroyed)
+      throw new errors.Error('ERR_HTTP2_INVALID_SESSION');
 
-    if (this[kHandle] !== undefined)
-      this[kHandle].destroying();
+    if (opaqueData !== undefined && !isArrayBufferView(opaqueData)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                  'opaqueData',
+                                  ['Buffer', 'TypedArray', 'DataView']);
+    }
+    if (typeof code !== 'number') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'code', 'number');
+    }
+    if (typeof lastStreamID !== 'number') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
+                                 'lastStreamID', 'number');
+    }
 
-    setImmediate(finishSessionDestroy.bind(this), socket);
+    const goawayFn = submitGoaway.bind(this, code, lastStreamID, opaqueData);
+    if (this.connecting) {
+      this.once('connect', goawayFn);
+      return;
+    }
+    goawayFn();
   }
 
-  // Graceful or immediate shutdown of the Http2Session. Graceful shutdown
-  // is only supported on the server-side
-  shutdown(options, callback) {
-    const state = this[kState];
-    if (state.destroyed || state.destroying)
-      throw new errors.Error('ERR_HTTP2_INVALID_SESSION');
-
-    if (state.shutdown || state.shuttingDown)
+  // Destroy the Http2Session, making it no longer usable and cancelling
+  // any pending activity.
+  destroy(error = NGHTTP2_NO_ERROR, code, lastStreamID, opaqueData) {
```

Comment:
We don't need the `lastStreamID`and `opaqueData` here.. I've removed those. The `code` is required because we have to know which error code to send in the final GOAWAY frame and that cannot always be determined automatically.